### DISCUSSION TITLE
Catalog request filterExpression

### DISCUSSION THREAD

**Discussion by Moritz72** [eclipse-edc/Connector#4253] [Q&A] [answered]
Hi,

I'm currently testing how the filterExpression for the catalog request works and have created an asset with this jsonLd:

```
{
  "@context": {},
  "@id": "100",
  "properties": {
    "name": "my_asset",
    "custom": {
        "customProperty": "test",
        "anotherProperty": "another-test"
    }
  },
  "dataAddress": {
    "@type": "DataAddress",
    "type": "AmazonS3",
    "keyName": "key",
    "region": "us-east-1",
    "bucketName": "bucket"
  }
}
```

According to [#3559](https://github.com/eclipse-edc/Connector/pull/3559) I thought that if I want to query for "customProperty" I should do

```
{
  "@context": {
    "edc": "https://w3id.org/edc/v0.0.1/ns/"
  },
  "@type": "CatalogRequest",
  "counterPartyAddress": "{{ALICE_CONTROL}}",
  "protocol": "dataspace-protocol-http",
  "querySpec": {
    "offset": 0,
    "limit": 50,
    "filterExpression": [
        {
            "operandLeft": "'https://w3id.org/edc/v0.0.1/ns/custom'.'https://w3id.org/edc/v0.0.1/ns/customProperty'",
            "operator": "=",
            "operandRight": "test"
        }
    ]
  }
}
```

But this does not work. Instead I have to replace "test" by "[{\"@value\":\"test\"}]". Is this intended or am I doing something wrong? I'm using EDC v0.5.1 in case that's relevant.

**Comment by paullatzelsperger** [comment, marked as answer]
You can take a look at [this test](https://github.com/eclipse-edc/Connector/blob/d22b18e193cee7d001069a47d703904a009cb23c/system-tests/management-api/management-api-test-runner/src/test/java/org/eclipse/edc/test/e2e/managementapi/AssetApiEndToEndTest.java#L218)

Also, `https://w3id.org/edc/v0.0.1/ns/customProperty` is incorrect, in an actual scenario you should be using a custom namespace for your custom properties. to do that, you'll need to modify the context.

**Reply by Moritz72** [reply]
Ok, thank you. I created an asset like in the test you linked and modified the  "operandLeft" accordingly and now it works.

Just as an aside when I was testing this with "/v3/assets/request" as it is in the test it didn't work because of the AssetQueryValidator.
```
{
  "@context": {
     "edc": "https://w3id.org/edc/v0.0.1/ns/"
  },
  "@type": "QuerySpec",
  "offset": 0,
  "limit": 50,
  "filterExpression": [
    {
      "operandLeft": "'https://w3id.org/edc/v0.0.1/ns/nested'.@id",
      "operator": "=",
      "operandRight": "1"
    }
  ]
}
```
returns
```
[
    {
        "message": "The query path must start with a letter or an '_' but was ''https://w3id.org/edc/v0.0.1/ns/nested'.@id'",
        "type": "InvalidRequest",
        "path": null,
        "invalidValue": null
    }
]
```

**Reply by ndr-brt** [reply]
this has been fixed *after* release 0.5.1, so please update your connector and the issue will be solved.

**Reply by Moritz72** [reply]
Ok, thank you.
